cmd/ocsp-responder/main.go

```diff
@@ -190,25 +210,163 @@ func (src *dbSource) Response(ctx context.Context, req *ocsp.Request) ([]byte, h
 		ctx, cancel = context.WithTimeout(ctx, src.timeout)
 		defer cancel()
 	}
-	certStatus, err = sa.SelectCertificateStatus(src.dbMap.WithContext(ctx), serialString)
-	if err != nil {
-		if db.IsNoRows(err) {
-			return nil, nil, bocsp.ErrNotFound
-		}
-		src.log.AuditErrf("Looking up OCSP response: %s", err)
-		return nil, nil, err
+
+	// The primary and secondary lookups send goroutines to get an OCSP
+	// status given a serial and return a channel of the output.
+	primaryChan := src.primaryLookup.getResponse(ctx, req)
+
+	// If the redis source is nil, don't try to get a response.
+	var secondaryChan chan lookupResponse
+	if src.secondaryLookup != nil {
+		secondaryChan = src.secondaryLookup.getResponse(ctx, req)
 	}
-	if certStatus.IsExpired {
-		src.log.Infof("OCSP Response not sent (expired) for CA=%s, Serial=%s", hex.EncodeToString(req.IssuerKeyHash), serialString)
-		return nil, nil, bocsp.ErrNotFound
-	} else if certStatus.OCSPLastUpdated.IsZero() {
-		src.log.Warningf("OCSP Response not sent (ocspLastUpdated is zero) for CA=%s, Serial=%s", hex.EncodeToString(req.IssuerKeyHash), serialString)
-		return nil, nil, bocsp.ErrNotFound
-	} else if !src.filter.responseMatchesIssuer(req, certStatus) {
-		src.log.Warningf("OCSP Response not sent (issuer and serial mismatch) for CA=%s, Serial=%s", hex.EncodeToString(req.IssuerKeyHash), serialString)
-		return nil, nil, bocsp.ErrNotFound
+
+	// If the primary source returns first, check the output and return
+	// it. If the secondary source wins, then wait for the primary so the
+	// results from the seconary can be verified. It is important that we
+	// never return a response from the redis source that is good if mysql
+	// has a revoked status. If the secondary source wins the race and
+	// passes these checks, return it's response instead.
+	select {
+	case <-ctx.Done():
+		err := fmt.Errorf("looking up OCSP response for serial: %s err: %w", serialString, ctx.Err())
+		src.log.Debugf(err.Error())
+		ocspLookups.WithLabelValues("canceled").Inc()
+		return nil, nil, err
+	case primaryResult := <-primaryChan:
+		if primaryResult.err != nil {
+			src.log.AuditErrf("Looking up OCSP response: %s", err)
+			ocspLookups.WithLabelValues("mysql_failed").Inc()
+			sourceUsed.WithLabelValues("error_returned").Inc()
+			return nil, nil, primaryResult.err
+		}
+		// Parse the OCSP bytes returned from the primary source to check
+		// status, expiration and other fields.
+		primaryParsed, err := ocsp.ParseResponse(primaryResult.bytes, nil)
+		if err != nil {
+			src.log.AuditErrf("parsing OCSP response: %s", err)
+			ocspLookups.WithLabelValues("mysql_failed").Inc()
```

Comment:
A parse error should be a different label; it's a much different (and more severe) category of error.

Reply:
Added labels for `mysql_parse_error` and `redis_parse_error` for these.